internal/graph/graph.go

```diff
@@ -0,0 +1,215 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package graph
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	errArgKind = errors.New("constructor arguments must be pointers")
+
+	_typeOfError = reflect.TypeOf((*error)(nil)).Elem()
+)
+
+// Graph contains all Graph for current graph
+type Graph struct {
+	sync.RWMutex
+
+	nodes map[interface{}]graphNode
+}
+
+// NewGraph creates new data Graph for dig
+func NewGraph() Graph {
+	return Graph{
+		nodes: make(map[interface{}]graphNode),
+	}
+}
+
+// Reset the graph
+func (g *Graph) Reset() {
+	g.Lock()
+	defer g.Unlock()
+	g.nodes = make(map[interface{}]graphNode)
+}
+
+// Read reads value from the Graph
+func (g *Graph) Read(objType reflect.Type) (reflect.Value, error) {
+	// check if the type is a registered objNode
+	g.RLock()
+	n, ok := g.nodes[objType]
+	g.RUnlock()
+	if !ok {
+		return reflect.Zero(objType), fmt.Errorf("type %v is not registered", objType)
+	}
+	v, err := n.value(g, objType)
+	if err != nil {
+		return reflect.Zero(objType), errors.Wrapf(err, "unable to resolve %v", objType)
+	}
+	return v, nil
+}
+
+// InsertObject the Graph with the provided value
+func (g *Graph) InsertObject(v reflect.Value, vtype reflect.Type) error {
+	g.Lock()
+	defer g.Unlock()
+	onode := objNode{
+		node: node{
+			objType:     vtype,
+			cached:      true,
+			cachedValue: v,
+		},
+	}
+	g.nodes[vtype] = &onode
+	return nil
+}
+
+// InsertConstructor adds the constructor to the Graph
+func (g *Graph) InsertConstructor(constr interface{}) error {
+	ctype := reflect.TypeOf(constr)
+	// count of number of objects to be registered from the list of return parameters
+	count := ctype.NumOut()
+	objTypes := make([]reflect.Type, count, count)
+	for i := 0; i < count; i++ {
+		objTypes[i] = ctype.Out(i)
+	}
+
+	nodes := make([]node, count, count)
+	for i := 0; i < count; i++ {
+		nodes[i] = node{
+			objType: objTypes[i],
+		}
+	}
+	argc := ctype.NumIn()
+	n := funcNode{
+		deps:        make([]interface{}, argc),
+		constructor: constr,
+		nodes:       nodes,
+	}
+	for i := 0; i < argc; i++ {
+		arg := ctype.In(i)
+		if arg.Kind() != reflect.Ptr && arg.Kind() != reflect.Interface {
```

Comment:
hmm, should the container care about it? Extending for slices and map will be hard then.